lightning-block-sync/src/http.rs

```diff
@@ -0,0 +1,765 @@
+		let mut message_length = HttpMessageLength::Empty;
+		loop {
+			let line = read_line!()
+				.ok_or(std::io::Error::new(std::io::ErrorKind::UnexpectedEof, "no headers"))?;
+			if line.is_empty() { break; }
+
+			let header = HttpHeader::parse(&line)?;
+			if header.has_name("Content-Length") {
+				let length = header.value.parse()
+					.map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
+				if let HttpMessageLength::Empty = message_length {
+					message_length = HttpMessageLength::ContentLength(length);
+				}
+				continue;
+			}
+
+			if header.has_name("Transfer-Encoding") {
+				message_length = HttpMessageLength::TransferEncoding(header.value.into());
+				continue;
+			}
+		}
+
+		if !status.is_ok() {
+			// TODO: Handle 3xx redirection responses.
+			return Err(std::io::Error::new(std::io::ErrorKind::NotFound, "not found"));
+		}
+
+		// Read message body
+		let read_limit = MAX_HTTP_MESSAGE_BODY_SIZE - reader.buffer().len();
+		reader.get_mut().set_limit(read_limit as u64);
+		match message_length {
+			HttpMessageLength::Empty => { Ok(Vec::new()) },
+			HttpMessageLength::ContentLength(length) => {
+				if length == 0 || length > MAX_HTTP_MESSAGE_BODY_SIZE {
+					Err(std::io::Error::new(std::io::ErrorKind::InvalidData, "out of range"))
+				} else {
+					let mut content = vec![0; length];
+					#[cfg(feature = "tokio")]
+					reader.read_exact(&mut content[..]).await?;
+					#[cfg(not(feature = "tokio"))]
+					reader.read_exact(&mut content[..])?;
+					Ok(content)
+				}
+			},
+			HttpMessageLength::TransferEncoding(coding) => {
+				if !coding.eq_ignore_ascii_case("chunked") {
+					Err(std::io::Error::new(
+							std::io::ErrorKind::InvalidInput, "unsupported transfer coding"))
+				} else {
+					let mut content = Vec::new();
+					#[cfg(feature = "tokio")]
+					{
+						// Since chunked_transfer doesn't have an async interface, only use it to
+						// determine the size of each chunk to read.
+						//
+						// TODO: Replace with an async interface when available.
+						// https://github.com/frewsxcv/rust-chunked-transfer/issues/7
+						loop {
+							// Read the chunk header which contains the chunk size.
+							let mut chunk_header = String::new();
+							reader.read_line(&mut chunk_header).await?;
+							if chunk_header == "0\r\n" {
+								// Read the terminator chunk since the decoder consumes the CRLF
+								// immediately when this chunk is encountered.
+								reader.read_line(&mut chunk_header).await?;
+							}
+
+							// Decode the chunk header to obtain the chunk size.
+							let mut buffer = Vec::new();
+							let mut decoder = chunked_transfer::Decoder::new(chunk_header.as_bytes());
+							decoder.read_to_end(&mut buffer)?;
+
+							// Read the chunk body.
+							let chunk_size = match decoder.remaining_chunks_size() {
+								None => break,
+								Some(chunk_size) => chunk_size,
+							};
+							let mut chunk_body = vec![0; chunk_size + "\r\n".len()];
```

Comment:
You can avoid the extra allocations by writing into content.
```
-                                                       let mut chunk_body = vec![0; chunk_size + "\r\n".len()];
-                                                       reader.read_exact(&mut chunk_body[..]).await?;
-                                                       content.extend_from_slice(&chunk_body[..chunk_size]);
+                                                       let orig_len = content.len();
+                                                       content.resize(orig_len + chunk_size + "\r\n".len(), 0);
+                                                       reader.read_exact(&mut content[orig_len..]).await?;
+                                                       content.resize(orig_len + chunk_size, 0);
```

Reply:
Done.